lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+                        mlir::isa<mlir::ViewLikeOpInterface>(op))
+                      continue;
+                    if (mlir::isa<mlir::memref::AllocOp>(op)) {
+                      gpuBufferAllocs.insert({op, {}});
+                    } else if (mlir::isa<mlir::func::CallOp>(op)) {
+                      // Ignore
+                    } else {
+                      op->emitError("Unhandled memref producer");
+                      return mlir::WalkResult::interrupt();
+                    }
+
+                  } else {
+                    auto block = alias.getParentBlock();
+                    auto blockArgs = block->getArguments();
+                    auto it = llvm::find(blockArgs, alias);
+                    assert(it != blockArgs.end());
+                    auto index = static_cast<unsigned>(it - blockArgs.begin());
+                    gpuBufferParams.insert({index, {}});
+                  }
+                }
+              }
+
+              return mlir::WalkResult::advance();
+            })
+            .wasInterrupted()) {
+      signalPassFailure();
+      return;
+    }
+
+    // Checks the access type of the OP
+    auto getAccessType = [&](mlir::Value memref) {
+      AccessType ret;
+      for (auto mem : aliases.resolve(memref)) {
+        for (auto user : mem.getUsers()) {
+          if (mlir::isa<mlir::func::ReturnOp>(user)) {
+            ret.hostRead = true;
+            ret.hostWrite = true;
+            continue;
+          }
+
+          if (auto copy = mlir::dyn_cast<mlir::memref::CopyOp>(user)) {
+            if (copy.source() == mem)
+              ret.hostRead = true;
+
+            if (copy.target() == mem)
+              ret.hostWrite = true;
+
+            continue;
+          }
+
+          if (auto memInterface =
+                  mlir::dyn_cast<mlir::MemoryEffectOpInterface>(user)) {
+            bool onDevice = user->getParentOfType<mlir::gpu::LaunchOp>();
+            if (memInterface.hasEffect<mlir::MemoryEffects::Read>())
+              (onDevice ? ret.deviceRead : ret.hostRead) = true;
+
+            if (memInterface.hasEffect<mlir::MemoryEffects::Write>())
+              (onDevice ? ret.deviceWrite : ret.hostWrite) = true;
+
+            continue;
+          }
+          if (mlir::isa<mlir::func::CallOp>(user)) {
+            bool onDevice = user->getParentOfType<mlir::gpu::LaunchOp>();
+            (onDevice ? ret.deviceRead : ret.hostRead) = true;
+            (onDevice ? ret.deviceWrite : ret.hostWrite) = true;
+            continue;
+          }
+        }
+      }
+      return ret;
+    };
+
+    for (auto &it : gpuBufferAllocs) {
+      auto alloc = mlir::cast<mlir::memref::AllocOp>(it.first);
+      it.second = getAccessType(alloc);
+    }
+
+    auto &block = funcBody.front();
+    for (auto &it : gpuBufferParams) {
+      auto param = block.getArgument(it.first);
+      it.second = getAccessType(param);
+
+      it.second.hostRead = true;
```

Comment:
Does this mean that the function always assumes the input/output is in the host side?   This is not TRUE for XLA usage, where the caller controls the runtime resource, so the input/output passed would be in device side for most of time. 

Are there a use case where there are two consecutive function calls, the first one produce a buffer for the second one to use, can the intermediate buffer stay in the device memory?  How the current code would be restructured to support that need?  

I don't mean that this PR needs to fix these limitations.  But would be useful if we list these considerations as future development opportunities.

Reply:
yes, you are right, the function assumes that the input/output is in host side. we dont have a test case of the above mentioned case. maybe in the future I will do a subsequent PR to address this?